src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -45,19 +45,173 @@ struct extent_to_write_t {
+	       << " pin_end " << info.pin_end
+	       << " left_paddr " << info.left_paddr
+	       << " right_paddr " << info.right_paddr
+	       << " data_begin " << info.data_begin
+	       << " data_end " << info.data_end
+	       << " aligned_begin " << info.aligned_begin
+	       << " aligned_end " << info.aligned_end
+	       << " block_size " << info.block_size
+	       << " data_size " << info.data_size
+	       << " write_size " << info.write_size << std::boolalpha
+	       << " split_left " << info.split_left
+	       << " split_right " << info.split_right << std::noboolalpha;
+  }
+  void init(laddr_t offset,
+	    extent_len_t len,
+	    const lba_pin_list_t& pins,
+	    extent_len_t _block_size) {
+    pin_begin = pins.front()->get_key();
+    pin_end = pins.back()->get_key() + pins.back()->get_length();
+    left_paddr = pins.front()->get_val();
+    right_paddr = pins.back()->get_val();
+    data_begin = offset;
+    data_end = offset + len;
+    aligned_begin = p2align((uint64_t)data_begin, (uint64_t)_block_size);
+    aligned_end = p2roundup((uint64_t)data_end, (uint64_t)_block_size);
+    block_size = _block_size;
+    data_size = aligned_end - aligned_begin;
+    write_size = pin_end - pin_begin;
+    split_left = false;
+    split_right = false;
+    check();
+    split();
+  }
+  void check() const {
+    ceph_assert(pin_begin <= data_begin);
+    ceph_assert(pin_end >= data_end);
+    ceph_assert(aligned_begin % block_size == 0);
+    ceph_assert(aligned_begin <= data_begin);
+    ceph_assert(pin_begin <= aligned_begin);
+    ceph_assert(aligned_end % block_size == 0);
+    ceph_assert(aligned_end >= data_end);
+    ceph_assert(pin_end >= aligned_end);
+  }
+  /*
+   * When trying to modify a portion of an object data block, follow
+   * the read-full-extent-then-merge-new-data strategy, if the write
+   * amplification caused by it is not greater than
+   * seastore_obj_data_write_amplification; otherwise, split the
+   * original extent into at most three parts: origin-left, part-to-be-modified
+   * and origin-right.
+   *
+   * all used variable are described as follows:
+   * |<--------------------------------------write_size--------------------------------------->|
+   *                 |<----------------------data_size----------------------->|
+   *                                data_begin         data_end
+   *                                ---------------------------(incoming data)
+   *                 aligned_begin                                  aligned_end
+   *                 ----------------------------------------------------------(incoming data's block-size-aligned range)
+   * |<--left_size-->|                                                        |<--right_size-->|
+   * pin_begin                                                                           pin_end
+   * -------------------------------------(existing extent)   ----------------------------------(existing extent)
```

Comment:
I don't understand this line.

Reply:
"existing extent" represents the pin_list returned from `Transaction::get_pins`
```
    pin_begin = pins.front()->get_key();
    pin_end = pins.back()->get_key() + pins.back()->get_length();
```